### DISCUSSION TITLE
Getting obscure error during SSR after upgrading Next13 to Next14

### DISCUSSION THREAD

**Discussion by pdubb** [vercel/next.js#73870] [Help]
### Summary

After upgrading Next from 13.5.6 to 14.2.20, we are getting the following error during SSR. 
```
Internal error: Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.
```

The browser shows a 500 error in the network tab, but once the client hydrates, everything renders as it should. Can someone offer any guidance as to why we're seeing this error? 

We have a complex app that was working fine on the previous version, and as I try to diagnose the issues I can't seem to pinpoint any one particular import causing the problem. I've run circular dependency checks to see if that might be a cause, but there are no issues there. 

Was something introduced in Next 14 that would cause something like this?

**Comment by mcollier** [comment]
@pdubb Did you ever figure this out?

**Reply by pdubb** [reply]
@mcollier It was a while ago, so I can't say what fixed it exactly, but we were able to get past it. If I recall, it required starting from scratch and updating one package at a time to see what kind of effect it had.

**Comment by naveen-kumar-b1** [comment]
## Follow-up: Make the webhook endpoint generic

The initial implementation had the orchestrator's `/webhooks/task` endpoint hardcoded for openhoof's event types (`task.started`, `step.completed`, etc.) with openhoof-specific formatting (step role headers, cost, truncation to 30KB for Jira). This doesn't scale — every new agent with long-running operations would need its own formatting logic in the orchestrator.

### Proposal: Generic event schema

Instead of the orchestrator knowing about each agent's event structure, make the webhook contract agent-agnostic:

```json
{
  "agent": "OpenHoofAgent",
  "event_type": "task.completed",
  "summary": "Task completed — PR created",
  "detail": "Added TODO section to README with 9 items. Security scan passed.",
  "url": "https://github.com/org/repo/pull/20",
  "metadata": {
    "cost_usd": 0.14,
    "duration_seconds": 97
  },
  "context": {}
}
```

Key differences:
- **`summary`** and **`detail`** are pre-formatted by the agent — the orchestrator doesn't need to know how to format step roles, models, or costs
- **`url`** is an optional link (PR, dashboard, log) that channels can render appropriately
- **`metadata`** is structured data for filtering/routing (e.g. only notify Slack on failures)
- The orchestrator's job is purely routing (context → channels) and delivery, not formatting

### What changes

| Concern | Original design | Generic design |
|---|---|---|
| Event schema | Openhoof-specific (step.role, step.model) | Agent-agnostic (summary, detail, url) |
| Formatting | Orchestrator formats per channel | Agent formats its own summary/detail |
| New agent | Add formatting logic to orchestrator | Agent provides summary — orchestrator just routes |
| Slack rendering | Orchestrator builds mrkdwn | Orchestrator wraps summary in a simple template |
| Jira rendering | Orchestrator reproduces openhoof's format | Orchestrator posts detail as-is |

### Interim: Scheduler approach is working

While we finalize this design, the system prompt change (use SchedulerAgent for polling instead of LLM turns) is deployed and working. Tasks complete in 3 LLM turns, scheduler checks status after 90s and posts results to the Slack thread via its built-in `notification_channel`/`notification_thread` support.

The webhook approach is still the better long-term architecture (event-driven vs polling), but making it generic from the start avoids coupling the orchestrator to individual agent implementations.
